tests/unit/query_graph/util.py

```diff
@@ -36,3 +40,41 @@ def to_dict(obj, exclude=None, include=None):
     if hasattr(obj, "dict"):
         return to_dict(obj.dict(), exclude=exclude, include=include)
     return obj
+
+
+def evaluate_and_compare_odfv_and_udf_results(
+    input_map, odfv_expr, udf_expr, odfv_stats=None, udf_stats=None, expected_output=None
+):
+    """Evaluate the odfv & udf expressions & compare the results"""
+    # check the odfv expression can be evaluated
+    locals = {"np": np, "pd": pd, "sp": sp, **input_map}
+    if odfv_stats:
+        for stat in odfv_stats:
+            if isinstance(stat, tuple):
+                stat = f"{stat[0]} = {stat[1]}"
+            exec(stat, locals)
+
+    out_odfv = eval(odfv_expr, locals)
+
+    # check the udf expression can be evaluated
+    locals = {"np": np, "pd": pd, "sp": sp}
+    if udf_stats:
+        for stat in udf_stats:
+            if isinstance(stat, tuple):
+                stat = f"{stat[0]} = {stat[1]}"
+            exec(stat, locals)
+
+    out_vals = []
+    input_param_names = list(input_map.keys())
+    for input_vals in zip(*input_map.values()):
+        for param_name, param_val in zip(input_param_names, input_vals):
+            locals[param_name] = param_val
+        out_vals.append(eval(udf_expr, locals))
+    out_udf = pd.Series(out_vals)
+
+    # check the consistency between two expressions
+    pd.testing.assert_series_equal(out_odfv, out_udf)
+
+    # check the expected output
+    if expected_output is not None:
+        pd.testing.assert_series_equal(out_odfv, expected_output)
```

Comment:
This is the key helper function that is used to evaluate on demand feature view & user defined function & compare their output.